assets/src/loader.rs

```diff
@@ -0,0 +1,348 @@
+    {
+        let inner = AssetFutureCell { value: UnsafeCell::new(None) };
+        let inner = Arc::new(inner);
+
+        let cloned = inner.clone();
+
+        pool.spawn(move || {
+            let res = f();
+            // This is safe because the other reference to `value` is only accessed
+            // if there is just one strong reference of the `Arc`. However, this closure holds
+            // one.
+            unsafe {
+                *cloned.value.get() = Some(res);
+            }
+
+            // Now, `cloned` gets dropped.
+        });
+
+        let inner = Some(inner);
+
+        AssetFuture { inner }
+    }
+}
+
+impl<A, E> Component for AssetFuture<A, E>
+    where A: Component,
+          E: 'static
+{
+    type Storage = DenseVecStorage<Self>;
+}
+
+impl<A, E> Future for AssetFuture<A, E> {
+    type Item = A;
+    type Error = E;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        use futures::Async;
+
+        match Arc::try_unwrap(self.inner.take().unwrap()) {
+            Ok(x) => {
+                // As soon as the worker thread finished executing the closure, the second
+                // reference gets dropped and we enter this branch.
+                // Thus, we are the only one with access to the cell.
+                unsafe {
+                    x.value
+                        .into_inner()
+                        .expect("Thread panicked")
+                        .map(Async::Ready)
+                }
+            }
+            Err(arc) => {
+                self.inner = Some(arc);
+
+                Ok(Async::NotReady)
+            }
+        }
+    }
+
+    fn wait(mut self) -> Result<Self::Item, Self::Error> {
+        use futures::Async;
+
+        loop {
+            match self.poll() {
+                Ok(Async::Ready(x)) => return Ok(x),
+                Ok(Async::NotReady) => {}
+                Err(x) => return Err(x),
+            }
+        }
+    }
+}
+
+struct AssetFutureCell<A, E> {
+    value: UnsafeCell<Option<Result<A, E>>>,
+}
+
+unsafe impl<A, E> Send for AssetFutureCell<A, E> {}
+unsafe impl<A, E> Sync for AssetFutureCell<A, E> {}
+
+/// The asset loader, holding the contexts,
+/// the default (directory) store and a reference to the
+/// `ThreadPool`.
+pub struct Loader {
+    contexts: FnvHashMap<TypeId, Box<Any>>,
+    directory: Arc<Box<AnyStore>>,
+    pool: Arc<ThreadPool>,
+    stores: FnvHashMap<String, Arc<Box<AnyStore>>>,
+}
+
+impl Loader {
+    /// Creates a new asset loader, initializing the directory store with the
+    /// given path.
+    pub fn new<P>(alloc: &Allocator, directory: P, pool: Arc<ThreadPool>) -> Self
+        where P: Into<PathBuf>
+    {
+        Loader {
+            contexts: Default::default(),
+            directory: Arc::new(Box::new(Directory::new(alloc, directory))),
+            pool: pool,
+            stores: Default::default(),
+        }
+    }
+
+    pub fn add_store<I, S>(&mut self, name: I, store: S)
```

Comment:
Oh yeah, probably, but I'd like every item in the crate to have docs. Should enable that lint.